### PR TITLE
HHH-1152 Discriminator based inheritance for embeddable types

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-1152

New version of https://github.com/hibernate/hibernate-orm/pull/8344 after first review iteration.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 1]
```diff
@@ -693,6 +814,60 @@ public <T> DomainResult<T> createDomainResult(
...
+			else {
+				results[i] = null;
+			}
+		}
+		return results;
+	}
+
+	@Override
+	protected void setAttributeValues(Object component, Object[] values) {
+		final String compositeClassName = component.getClass().getName();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:856]: Maybe it's worth delegating to `super.setAttributeValues()` if the type is not polymorphic?

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 2]
```diff
@@ -693,6 +814,60 @@ public <T> DomainResult<T> createDomainResult(
...
+		final AttributeMapping attributeMapping = getAttributeMapping( position );
+		if ( declaresAttribute( instance.getClass().getName(), attributeMapping ) ) {
+			return attributeMapping.getValue( instance );
+		}
+		return null;
+	}
+
+	@Override
+	protected Object[] getAttributeValues(Object compositeInstance) {
+		final Object[] results = new Object[getNumberOfAttributeMappings()];
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:837]: Maybe it's worth delegating to `super.getAttributeValues()` if the type is not polymorphic?

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java [hunk 3]
```diff
@@ -103,14 +108,35 @@ default boolean shouldSkipInitializer(RowProcessingState rowProcessingState) {
...
+						.getEmbeddableTypeDescriptor();
+				if ( descriptor.isPolymorphic() ) {
+					// The embeddable is polymorphic, check if the current subtype defines the initialized attribute
+					final AttributeMapping attribute = getInitializedPart().asAttributeMapping();
+					if ( attribute != null ) {
+						embeddableInitializer.resolveKey( rowProcessingState );
+						final String embeddableClassName = descriptor.getDiscriminatorMapping()
+								.resolveDiscriminatorValue( embeddableInitializer.getDiscriminatorValue() )
+								.getIndicatedEntityName();
+						embeddableInitializer.getDiscriminatorValue();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/FetchParentAccess.java:125]: Unnecessary call?

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java [hunk 4]
```diff
@@ -91,8 +92,12 @@ public Object[] getValues(Object compositeInstance) {
 			return optimizer.getAccessOptimizer().getPropertyValues( compositeInstance );
 		}
 
+		return getAttributeValues( compositeInstance );
+	}
+
+	protected Object[] getAttributeValues(Object compositeInstance) {
 		final Object[] results = new Object[getNumberOfAttributeMappings()];
-		for ( int i = 0; i < results.length; i++ ) {
+		for ( int i = 0; i < getNumberOfAttributeMappings(); i++ ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:100]: Unnecessary change?

```suggestion
		for ( int i = 0; i < results.length; i++ ) {
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java [hunk 5]
```diff
@@ -91,8 +92,12 @@ public Object[] getValues(Object compositeInstance) {
 			return optimizer.getAccessOptimizer().getPropertyValues( compositeInstance );
 		}
 
+		return getAttributeValues( compositeInstance );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:95]: What do you think about adding the class name as string as last value to this array if the type is polymorphic, similar to what happens for `CallableStatement` handling?

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java [hunk 6]
```diff
@@ -108,9 +113,13 @@ public void setValues(Object component, Object[] values) {
 			optimizer.getAccessOptimizer().setPropertyValues( component, values );
 		}
 		else {
-			for ( int i = 0; i < values.length; i++ ) {
-				getAttributeMapping( i ).getPropertyAccess().getSetter().set( component, values[i] );
-			}
+			setAttributeValues( component, values );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:116]: If we add the class name as last value, this method should check that the instance is of the correct type I guess.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java [hunk 7]
```diff
@@ -91,8 +92,12 @@ public Object[] getValues(Object compositeInstance) {
 			return optimizer.getAccessOptimizer().getPropertyValues( compositeInstance );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:92]: The generated access optimizer implementation should also be adapted to add the class name to the end if we decide to do this.

File: hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java [hunk 8]
```diff
@@ -115,20 +114,58 @@ public static Object[] getJdbcValues(
...
+		assert jdbcIndex == valueCount;
 		return jdbcValues;
 	}
 
+	public static Object[] getValues(EmbeddableMappingType embeddableMappingType, Object domainValue) {
+		final Object[] attributeValues = embeddableMappingType.getValues( domainValue );
+		if ( embeddableMappingType.isPolymorphic() ) {
+			final Object[] result = new Object[attributeValues.length + 1];
+			System.arraycopy( attributeValues, 0, result, 0, attributeValues.length );
+			result[attributeValues.length] = domainValue.getClass().getName();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java:134]: This "hack" could go away if we decide to add the FQN as last attribute value to the array.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java (hunk 6 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:116]: If we did this we would need to convert the discriminator value in `AbstractEmbeddableInitializer` to the class name though, which we avoided in favor of accessing the instantiator directly by object.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:116]: I guess you're right and this method shouldn't do this check then to avoid the map lookup.

File: hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java [hunk 9]
```diff
@@ -115,20 +114,58 @@ public static Object[] getJdbcValues(
...
 					jdbcIndex,
 					options
 			);
 		}
-		assert jdbcIndex == jdbcValueCount;
+		assert jdbcIndex == valueCount;
 		return jdbcValues;
 	}
 
+	public static Object[] getValues(EmbeddableMappingType embeddableMappingType, Object domainValue) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java:129]: Should go away.

File: hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java [hunk 10]
```diff
@@ -1104,22 +1101,21 @@ private static Object fromString(
 			default:
 				if ( jdbcMapping.getJdbcType() instanceof AggregateJdbcType ) {
 					final AggregateJdbcType aggregateJdbcType = (AggregateJdbcType) jdbcMapping.getJdbcType();
-					final Object[] subValues = aggregateJdbcType.extractJdbcValues(
-							CharSequenceHelper.subSequence(
-									string,
-									start,
-									end
+					final StructAttributeValues subValues = StructHelper.getAttributeValues(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java:1104]: This change looks wrong to me i.e. the `getAttributeValues` call should be moved into the `if ( returnEmbeddable )` branch.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java (hunk 7 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java:92]: Will be done as part of https://hibernate.atlassian.net/browse/HHH-18110

File: hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java (hunk 9 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java]: Forgot, thanks.

File: hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java (hunk 10 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java]: Done.

**Review by beikov — APPROVED** [review]
